### PR TITLE
MGMT-4715 Auto install Day 2 hosts in kube-api

### PR DESCRIPTION
**Pull request by rollandf**
In case that the Cluster is a Day 2 cluster, hosts will be automatically installed if they are in Known status and Approved

### CODE REVIEW

**Comment by rollandf** [comment]
/hold
Need to test e2e

File: internal/bminventory/inventory.go [hunk 1]
```diff
@@ -1254,6 +1255,76 @@ func (b *bareMetalInventory) InstallClusterInternal(ctx context.Context, params
...
+
+	if cluster, err = common.GetClusterFromDB(b.db, clusterId, common.UseEagerLoading); err != nil {
+		return err
+	}
+	if h, err = b.getHost(ctx, clusterId.String(), hostId.String()); err != nil {
+		return err
+	}
+	// auto select host roles if not selected yet.
+	err = b.db.Transaction(func(tx *gorm.DB) error {
+		if err = b.hostApi.AutoAssignRole(ctx, h, tx); err != nil {
```
Comment (tsorya) [inline internal/bminventory/inventory.go:1273]: host role in day2?

File: internal/bminventory/inventory.go [hunk 2]
```diff
@@ -1254,6 +1255,76 @@ func (b *bareMetalInventory) InstallClusterInternal(ctx context.Context, params
 	return cluster, nil
 }
 
+func (b *bareMetalInventory) InstallSingleDay2HostInternal(ctx context.Context, clusterId strfmt.UUID, hostId strfmt.UUID) error {
```
Comment (eranco74) [inline internal/bminventory/inventory.go:1258]: I guess you can refactor the current `InstallSingleDay2Host` implementation to use this, no?

File: internal/controller/controllers/clusterdeployments_controller.go [hunk 3]
```diff
@@ -143,24 +143,52 @@ func (r *ClusterDeploymentsReconciler) Reconcile(req ctrl.Request) (ctrl.Result,
...
+			r.Log.Infof("Installing clusterDeployment %s %s", clusterDeployment.Name, clusterDeployment.Namespace)
+			var ic *common.Cluster
+			ic, err = r.Installer.InstallClusterInternal(ctx, installer.InstallClusterParams{
+				ClusterID: *cluster.ID,
+			})
+			if err != nil {
+				return r.updateState(ctx, clusterDeployment, cluster, err)
+			}
+			return r.updateState(ctx, clusterDeployment, ic, nil)
+		}
```
Comment (eranco74) [inline internal/controller/controllers/clusterdeployments_controller.go:162]: Consider extracting to another function

File: internal/bminventory/inventory.go (hunk 2 above)
Comment (rollandf) [reply internal/bminventory/inventory.go:1258]: There are several transactions and loops in `InstallHosts` and it makes it hard to refactor the function to work on a single host.
Can you look at `InstallHosts` and tell me if it worth the effort?

File: internal/bminventory/inventory.go [hunk 4]
```diff
@@ -1254,6 +1255,76 @@ func (b *bareMetalInventory) InstallClusterInternal(ctx context.Context, params
...
+			return err
+		}
+		return nil
+	})
+
+	if err != nil {
+		return err
+	}
+
+	if err = b.refreshAllHosts(ctx, cluster); err != nil {
```
Comment (filanov) [inline internal/bminventory/inventory.go:1283]: why do you need to call refresh all hosts? maybe just the relevant host?

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 3 above)
Comment (filanov) [reply internal/controller/controllers/clusterdeployments_controller.go:162]: +1

File: internal/bminventory/inventory.go (hunk 4 above)
Comment (tsorya) [reply internal/bminventory/inventory.go:1283]: maybe we can remove transaction from it? it should set only "worker role" in auto assign part. Does it worth all that logic?

File: internal/controller/controllers/clusterdeployments_controller.go (hunk 3 above)
Comment (rollandf) [reply internal/controller/controllers/clusterdeployments_controller.go]: Done

File: internal/bminventory/inventory.go (hunk 4 above)
Comment (rollandf) [reply internal/bminventory/inventory.go]: I refreshed only the specific host.

File: internal/bminventory/inventory.go [hunk 5]
```diff
@@ -1254,6 +1255,79 @@ func (b *bareMetalInventory) InstallClusterInternal(ctx context.Context, params
...
+		return err
+	}
+	// auto select host roles if not selected yet.
+	err = b.db.Transaction(func(tx *gorm.DB) error {
+		if err = b.hostApi.AutoAssignRole(ctx, h, tx); err != nil {
+			return err
+		}
+		return nil
+	})
+
```
Comment (tsorya) [inline internal/bminventory/inventory.go:1278]: remove empty line

Comment (rollandf) [reply internal/bminventory/inventory.go]: Done

**Review by filanov — APPROVED** [review]

**Comment by alknopfler** [comment]
Just  a question, doc for the CRD should be useful @rollandf

**Comment by rollandf** [comment]
> Just a suggestion, Doc for the CRD should be useful @rollandf

@alknopfler On which CRD?

**Comment by alknopfler** [comment]
My fault! Forget it ;)

**Pull request merged**